executor/src/witgen/global_constraints.rs

```diff
@@ -268,8 +266,52 @@ fn propagate_constraints<T: FieldElement>(
     (known_constraints, remove)
 }
 
+pub fn get_roots<T: FieldElement>(
+    expr: &Expression<T>,
+) -> Option<Vec<crate::witgen::EvalValue<&AlgebraicReference, T>>> {
+    match expr {
+        Expression::BinaryOperation(AlgebraicBinaryOperation {
+            left,
+            op: AlgebraicBinaryOperator::Mul,
+            right,
+        }) => {
+            let right_root = get_roots(right)?;
+            let mut left_roots = get_roots(left)?;
+
+            if !left_roots.is_empty() && !right_root.is_empty() {
+                if let ([(id1, Constraint::Assignment(_))], [(id2, Constraint::Assignment(_))]) = (
+                    &left_roots[0].constraints[..],
+                    &right_root[0].constraints[..],
+                ) {
+                    if id1 != id2 {
+                        return None;
+                    }
+                }
+            }
+
+            left_roots.extend(right_root);
+            Some(left_roots)
+        }
+        Expression::Number(_) => Some(vec![]),
+        _ => {
+            let symbolic_ev = SymbolicEvaluator;
+            let _root = ExpressionEvaluator::new(symbolic_ev)
```

Comment:
Please don't user underscores like that, we won't get "unused" warnings if you do it.